Updated German support for ordinals (+threshold) and fixed bugs

fquirin opened the pull request:
- Updated Readme (including ordinals_threshold)
- Added ordinals support for German
- Fixed several bugs in German parser and improved code
- Added more tests

fquirin commented:
@rtxm I've made some larger changes to 'parsers' and 'transforms' but it's almost completely related to German. I think we should split these files at some point into language dependent classes with interfaces (as done with 'Language').

I'm not sure why circleci complaints about 1 failing check, I can't see any problems in my code :thinking:

rtxm commented:
Thank you. The CI is complaining about PEP8 : run `flake8 --max-line-length 120 text_to_num` to check them.

rtxm commented on README.rst:
```diff
@@ -128,23 +128,27 @@ French:
+    >>> sentence = "Cinquième premier second troisième vingt et unième centième mille deux cent trentième."
+    >>> print(alpha2digit(sentence, "fr", ordinal_threshold=3))
+    5ème premier second troisième 21ème 100ème 1230ème.
+
```
Thank you for spotting this!

rtxm commented on tests/test_text_to_num_en.py:
```diff
@@ -132,13 +132,16 @@ def test_alpha2digit_ordinals(self):
+        source = "twenty second position at the twenty first event lost after the first second"
+        expected = "22nd position at the 21st event lost after the first second"
+        self.assertEqual(alpha2digit(source, "en"), expected)
+
```
Thanks for improving coverage!

rtxm reviewed (commented):
Thanks for this contribution. The German algorithm has significantly diverged from the base one and became quite complicated. I'm not sure it was a good idea to let the first contributor start from scratch instead of leveraging the existing base — in particular the whole `equation` thing based on `eval` bothers me and shouldn't be necessary IMHO. What do you think?

Anyway, for this beta support, correctness and accuracy, along with good test coverage are the most important so we can safely refactor in the future. You made a great work here!

fquirin commented:
> The German algorithm has significantly diverged from the base one and became quite complicated. I'm not sure it was a good idea to let the first contributor start from scratch instead of leveraging the existing base — in particular the whole equation thing based on eval bothers me and shouldn't be necessary IMHO. What do you think?

I agree. I think it is possible to refactor the code to better fit into the original interface, but there are a couple of things that are hard to manage. Especially the fact that German numbers are usually written as "eintausendsiebenhundertzweiundneunzig" but hardly any ASR software (I'm using it for speech recognition) will ever produce this output format adds a whole level of complexity. I think the split function should be called only once for each sentence but it is designed to split-or-fail (and converts everything to lower-case) so this requires a number of optimizations.
The "eval" construct is indeed very strange and a bit tooo convenient :sweat_smile: , but I think this is rather easy to replace :thinking: ... with a bit of effort.

All in all its a good baseline and we have enough tests now to evaluate new code in terms of coverage and speed ^^.

> The CI is complaining about PEP8 : run flake8 --max-line-length 120 text_to_num to check them.

I'll check that and make another commit tomorrow :-)

fquirin commented:
@rtxm I've updated German to handle one more edge case, removed some 'eval' sections that were trivial to remove (unless I'm missing something here) and did some code cleanups (long lines etc.).
Can't seem to find the remaining failed checks. My linter doesn't show anything anymore, at least nothing that wasn't previously there already like 1/2-letter or capital case variables and formatting of test code.

rtxm commented:
Thanks. Here are the `flake8` complains:
```
text_to_num/parsers.py:241:9: E265 block comment should start with '# '
text_to_num/parsers.py:282:9: E265 block comment should start with '# '
text_to_num/parsers.py:382:46: E221 multiple spaces before operator
text_to_num/parsers.py:504:13: E265 block comment should start with '# '
text_to_num/parsers.py:505:13: E265 block comment should start with '# '
text_to_num/transforms.py:77:38: E261 at least two spaces before inline comment
text_to_num/transforms.py:274:48: E261 at least two spaces before inline comment
text_to_num/lang/german.py:170:57: W605 invalid escape sequence '\s'
text_to_num/lang/german.py:170:63: E261 at least two spaces before inline comment
text_to_num/lang/german.py:221:43: W291 trailing whitespace
text_to_num/lang/german.py:242:28: E261 at least two spaces before inline comment
text_to_num/lang/german.py:268:21: E265 block comment should start with '# '
```
